src/V3Fork.cpp

```diff
@@ -34,12 +46,351 @@
+                varp = varp->cloneTree(false);
+                varp->direction(VDirection::NONE);
+            } else {
+                varp->unlinkFrBack();
+            }
+            varp->funcLocal(false);
+            varp->varType(VVarType::MEMBER);
+            varp->lifetime(VLifetime::AUTOMATIC);
+            varp->usedLoopIdx(false);  // No longer unrollable
+            m_instance.classp->addStmtsp(varp);
+        }
+
+        // Create class's constructor
+        AstFunc* const newp = new AstFunc{m_instance.classp->fileline(), "new", nullptr, nullptr};
+        newp->isConstructor(true);
+        newp->classMethod(true);
+        newp->dtypep(newp->findVoidDType());
+        m_instance.classp->addStmtsp(newp);
+    }
+
+    void linkNodes(VMemberMap& memberMap) {
+        UASSERT_OBJ(m_instance, m_procp, "No dynamic scope prototype");
+        UASSERT_OBJ(!linked(), m_instance.handlep, "Handle already linked");
+
+        AstNode* stmtp = getProcStmts();
+        UASSERT(stmtp, "trying to instantiate dynamic scope while not under proc");
+        VNRelinker stmtpHandle;
+        stmtp->unlinkFrBackWithNext(&stmtpHandle);
+
+        // Find node after last variable declaration
+        AstNode* initp = stmtp;
+        while (initp && VN_IS(initp, Var)) initp = initp->nextp();
+        UASSERT(stmtp, "Procedure lacks body");
+        UASSERT(initp, "Procedure lacks statements besides declarations");
+
+        AstNew* newp = new AstNew{m_procp->fileline(), nullptr};
+        newp->taskp(VN_AS(memberMap.findMember(m_instance.classp, "new"), NodeFTask));
+        newp->dtypep(m_instance.refDTypep);
+        newp->classOrPackagep(m_instance.classp);
+
+        AstNode* asgnp = new AstAssign{
+            m_procp->fileline(),
+            new AstVarRef{m_procp->fileline(), m_instance.handlep, VAccess::WRITE}, newp};
+
+        AstNode* inits = nullptr;  // Arguments need to be copied
+        for (AstVar* varp : m_captures) {
+            if (VL_LIKELY(varp->direction() != VDirection::INPUT)) continue;
+
+            AstMemberSel* const membersel = new AstMemberSel{
+                varp->fileline(),
+                new AstVarRef{varp->fileline(), m_instance.handlep, VAccess::WRITE},
+                varp->dtypep()};
+            membersel->name(varp->name());
+            membersel->varp(VN_AS(memberMap.findMember(m_instance.classp, varp->name()), Var));
+            AstNode* initAsgnp = new AstAssign{
+                varp->fileline(), membersel, new AstVarRef{varp->fileline(), varp, VAccess::READ}};
+            inits = AstNode::addNext(inits, initAsgnp);
+        }
+        if (inits) AstNode::addNext(asgnp, inits);
+
+        if (initp != stmtp) {
+            initp->addHereThisAsNext(asgnp);
+        } else {
+            AstNode::addNext(asgnp, static_cast<AstNode*>(initp));
+            stmtp = asgnp;
+        }
+
+        AstNode::addNext(static_cast<AstNode*>(m_instance.handlep), stmtp);
```

Comment:
Why is cast needed?

Reply:
```cxx
In file included from ../V3Fork.cpp:46:
../V3Ast.h: In instantiation of ‘static T_NodeResult* AstNode::addNext(T_NodeResult*, T_NodeNext*) [with T_NodeResult = AstVar; T_NodeNext = AstNode]’:
../V3Fork.cpp:167:25:   required from here
../V3Ast.h:1913:66: error: static assertion failed: 'T_NodeNext' must be a subtype of 'T_NodeResult'
 1913 |         static_assert(std::is_base_of<T_NodeResult, T_NodeNext>::value,
      |                                                                  ^~~~~
../V3Ast.h:1913:66: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
```

Because that's what happens if there's no such cast.